### PR TITLE
8267418: IntelliJ build and test of JavaFX does not work

### PR DESCRIPTION
**Pull request by Maran23**
~~Question: I was wondering, should I create a ticket for this as well? Given the fact that I don't have an https://bugs.openjdk.java.net account, I need to use the official bug reporting tool, which looked a bit overkill to me since someone needs to check my created ticket, while this PR is only affecting the IntelliJ IDE with OpenJFX and not the JavaFX platform directly.~~
EDIT: Thank you, Kevin. :)

This PR fixes the errors you get when cloning and working with OpenJFX in IntelliJ IDE:
- The **.idea/misc.xml** is modified to use **JDK_11** as language level instead of JDK_8. 
-> This is the language level shown inside the **Project Structure**. (File -> Project Structure...)
- The **.idea/base.iml, .idea/controls.iml, .idea/fxml.iml, .idea/web.iml, .idea/graphics.iml** are modified to include/recognize the shims (as test resource, this is very similar to the configuration inside the .classpath file from Eclipse)
- EDIT: The projects are now recognized by IntelliJ-gradle (**.idea/gradle.xml**, **.idea/compiler.xml**)

**-> With this, I can run all normal tests with IntelliJ**

### What I couldn't fix yet (When I tried, it looked like IntelliJ is overriding the settings on next gradle reload): 
- IntelliJ is not detecting javafx.graphic inside the shims
- All javafx.* dependencies are not found for the system tests

**-> If someone has a solution, feel free to comment :)**

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267418](https://bugs.openjdk.java.net/browse/JDK-8267418): IntelliJ build and test of JavaFX does not work

### Reviewers
 * [Ambarish Rapte](https://openjdk.java.net/census#arapte) (@arapte - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/506/head:pull/506` \
`$ git checkout pull/506`

Update a local copy of the PR: \
`$ git checkout pull/506` \
`$ git pull https://git.openjdk.java.net/jfx pull/506/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 506`

View PR using the GUI difftool: \
`$ git pr show -t 506`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/506.diff">https://git.openjdk.java.net/jfx/pull/506.diff</a>

</details>

### CODE REVIEW

**Comment by kevinrushforth** [comment]
> Question: I was wondering, should I create a ticket for this as well?

All fixes need a bug ID in JBS. For this one, it seemed easier for me to just file it directly, so I did: [JDK-8267418](https://bugs.openjdk.java.net/browse/JDK-8267418).

File: .idea/gradle.xml [hunk 1]
```diff
@@ -1,3 +1,28 @@
...
+            <option value="$PROJECT_DIR$/modules/javafx.swing" />
+            <option value="$PROJECT_DIR$/modules/javafx.swt" />
+            <option value="$PROJECT_DIR$/modules/javafx.web" />
+            <option value="$PROJECT_DIR$/tests/system" />
+          </set>
+        </option>
+      </GradleProjectSettings>
+    </option>
+  </component>
+</project>
```
Comment (arapte) [inline .idea/gradle.xml:28]: I can see the same modules in project structure with or without this change. Is this required for build to work ?

File: .idea/modules.xml [hunk 2]
```diff
@@ -25,5 +25,4 @@
       <module fileurl="file://$PROJECT_DIR$/.idea/web.iml" filepath="$PROJECT_DIR$/.idea/web.iml" />
     </modules>
   </component>
-</project>
-
+</project>
```
Comment (arapte) [inline .idea/modules.xml:28]: This change should be removed from PR.

**Review by arapte — COMMENTED** [review]
Is a specific version of IntelliJ required to test this fix, I use 2019.2 version. I always build using terminal, on my IntelliJ the build fails both before and after.

**Comment by Maran23** [comment]
> Is a specific version of IntelliJ required to test this fix, I use 2019.2 version. I always build using terminal, on my IntelliJ the build fails both before and after.

### I'm using the newest version (Community edition, 2021.1.2).

Executing `gradle` via the IntelliJ build-in terminal works for me:
image

Executing `gradle sdk` via 'IntelliJ Gradle' works for me:
image

And running normal unit tests works for me as well:
image

Any chance you can test it with the newest version?

File: .idea/gradle.xml (hunk 1 above)
Comment (Maran23) [reply .idea/gradle.xml:28]: This enabled the IntelliJ gradle plugin for me, so e.g. the support for doing the tasks in the ide. And with it, I was able to run normal unit tests from the ide (see also screenshot(s) below)
image

**Comment by arapte** [comment]
I tested with IntelliJ IDEA 2021.1.2 (Community Edition) Build #IC-211.7442.40, built on June 1, 2021.

I see no side effects of this change. However, I also don't see what you see too.
I tried it with a fresh clone of repository. 
The build from IntelliJ's inbuilt terminal worked without this change too, that terminal picks JDK/gradle/ant that was set in bashrc. 

I also don't see the modules in gradle window. Here is the screenshot that I see when the project(with this PR change) is loaded for first time in the IntelliJ.
image

Is it possible that some local changes on your machine/repo result in what you see?
Some idea files are ignored using .gitignore file, so they get modified they won't be shown in 'git status'
**from .gitignore file**
```
# Ignore IntelliJ files
.idea/tasks.xml
.idea/workspace.xml
.idea/inspectionProfiles/**
.idea/copyright/profiles_settings.xml
.idea/codeStyles/**
.idea/checkstyle-idea.xml
```
It would be helpful to compare those files before and after loading the project in IntelliJ.
Also, can you test the PR once, by cloning a fresh copy of repo and then apply the PR.

**Comment by Maran23** [comment]
> I tested with IntelliJ IDEA 2021.1.2 (Community Edition) Build #IC-211.7442.40, built on June 1, 2021.
> 
> I see no side effects of this change. However, I also don't see what you see too.
> I tried it with a fresh clone of repository.
> The build from IntelliJ's inbuilt terminal worked without this change too, that terminal picks JDK/gradle/ant that was set in bashrc.
> 
> I also don't see the modules in gradle window. Here is the screenshot that I see when the project(with this PR change) is loaded for first time in the IntelliJ.
> image

This might not work because of the dependency verification error you have in your console. I had this as well and I temporary deleted the **verification-metadata.xml** inside the **gradle** folder. As stated in the mailing list you can also delete the **.gradle** **cache** and **native** folder, but that's the easier way to test it though. Then you can click on the refresh icon in the gradle window and everything should build and then the projects should be visible in the gradle window.

**Comment by arapte** [comment]
> This might not work because of the dependency verification error you have in your console. I had this as well and I temporary deleted the **verification-metadata.xml** inside the **gradle** folder. 

Deleting **verification-metadata.xml**, solves the error on my system. The modules and tasks get listed in gradle window. Could execute graphics:test task from the gradle window.
The step to delete **verification-metadata.xml** seems acceptable. (May be we should include the whole IntelliJ setup steps with build documentation.)
It does not seem like a stopper to me, but would like to get @kevinrushforth 's opinion.

**Comment by arapte** [comment]
From the gradle documentation: [Skipping Javadocs and sources](https://docs.gradle.org/current/userguide/dependency_verification.html#sec:skipping-javadocs)
The verification of auto downloaded *-sources.jar files can be skipped using following change. This way we can avoid deleting the `verification-metadata.xml` file.

```
diff --git a/gradle/verification-metadata.xml b/gradle/verification-metadata.xml
index abacd0b05b..0a3d33726d 100644
--- a/gradle/verification-metadata.xml
+++ b/gradle/verification-metadata.xml
@@ -3,6 +3,9 @@
    <configuration>
       <verify-metadata>true</verify-metadata>
       <verify-signatures>false</verify-signatures>
+      <trusted-artifacts>
+         <trust file=".*-sources[.]jar" regex="true"/>
+      </trusted-artifacts>
    </configuration>
    <components>
       <component group="" name="ffmpeg-3.3.3" version="">

```

I tested this change in addition to the PR changes, dependency verification does not fail with this change but fails without.
Can you please verify and test this change preferably on a clean repo by creating a new IntelliJ project.

**Comment by kevinrushforth** [comment]
Good find. If this works, it seems like a good compromise to me. We don't download any `*-sources` or `*-javadoc` artifacts in our build, but it seems that IDEs do.

If we add this exclusion, I think we should do it using separate bug ID and a separate PR (requiring two reviewers, since this is an area we need to be careful about). As part of that, we would need to update the [`gradle/README.txt`](https://github.com/openjdk/jfx/blob/master/gradle/README.txt) file to let people know why the `*-sources` / `*-javadoc`  exclusion is there.

**Comment by Maran23** [comment]
> ```
> diff --git a/gradle/verification-metadata.xml b/gradle/verification-metadata.xml
> index abacd0b05b..0a3d33726d 100644
> --- a/gradle/verification-metadata.xml
> +++ b/gradle/verification-metadata.xml
> @@ -3,6 +3,9 @@
>     <configuration>
>        <verify-metadata>true</verify-metadata>
>        <verify-signatures>false</verify-signatures>
> +      <trusted-artifacts>
> +         <trust file=".*-sources[.]jar" regex="true"/>
> +      </trusted-artifacts>
>     </configuration>
>     <components>
>        <component group="" name="ffmpeg-3.3.3" version="">
> ```
> 
> I tested this change in addition to the PR changes, dependency verification does not fail with this change but fails without.
> Can you please verify and test this change preferably on a clean repo by creating a new IntelliJ project.

Good catch indeed! This works like a charm. I think this is a good follow-up.

**Comment by arapte** [comment]
The follow on is reported here: [JDK-8269244](https://bugs.openjdk.java.net/browse/JDK-8269244)

**Review by arapte — APPROVED** [review]
Approving, I re-verified the change: Loading project in IntelliJ does not cause any other changes in idea files.

**Pull request closed without merging**